### PR TITLE
[RW-11948][risk=low] upgrade springboot to 3, spring framework to 6, upgrade hibernate, etc.  

### PR DESCRIPTION
**Pull request by yonghaoy**
https://docs.google.com/document/d/1nJxR4y1coTY89gRfs163nUywjMRQqQqeLq0Wm6ub_SY/edit?resourcekey=0-g9gNQUaYUXjtIH9t5UtCIg&tab=t.0#heading=h.hxla6cwzss53

---
**PR checklist**

- [ ] I have included an issue ID or "no ticket" in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [ ] I have included a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [ ] I have manually tested this change and my testing process is described above.
- [ ] This change includes appropriate automated tests, and I have documented any behavior that cannot be tested with code.
- [ ] I have added explanatory comments where the logic is not obvious.
- One or more of the following is true:
  - [ ] This change is intended to complete a JIRA story, so I have checked that all AC are met for that story.
  - [ ] This change fixes a bug, so I have ensured the steps to reproduce are in the Jira ticket or provided above.
  - [ ] This change impacts deployment safety (e.g. removing/altering APIs which are in use), so I have documented the impacts in the description.
  - [ ] This change includes a new feature flag, so I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later.
  - [ ] This change modifies the UI, so I have taken screenshots or recordings of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P).
  - [ ] This change modifies API behavior, so I have run the relevant E2E tests locally because API changes are not covered by our PR checks.
  - [ ] None of the above apply to this change.

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/absorb/Credentials.java [hunk 1]
```diff
@@ -1,6 +1,6 @@
 package org.pmiops.workbench.absorb;
 
-import javax.annotation.Nullable;
+import jakarta.annotation.Nullable;
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/absorb/Credentials.java:3]: optional: can also change these to `org.jetbrains.annotations.Nullable;`

File: api/src/main/java/org/pmiops/workbench/api/UserController.java [hunk 2]
```diff
@@ -6,6 +6,8 @@
 import com.google.api.services.cloudbilling.model.ListBillingAccountsResponse;
 import com.google.common.base.Strings;
 import com.google.common.collect.Lists;
+import jakarta.annotation.Nonnull;
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/UserController.java:9]: optional: could use org.jetbrains.annotations.NotNull here instead

File: api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java [hunk 3]
```diff
@@ -391,14 +391,11 @@ public void testCreateUser_otherTextMustBeNullForNonOtherDiscoverySource() {
...
-    providedWorkbenchConfig.access.currentDuccVersions.forEach(
-        version -> {
-          DbUser user = userDao.findUserByUsername(USERNAME);
-          user.setDuccAgreement(signDucc(user, version));
-          user = userDao.save(user);
-          userService.syncDuccVersionStatus(user, Agent.asSystem());
-          verify(accessModuleService, never()).updateCompletionTime(any(), any(), any());
-        });
+    DbUser user = userDao.findUserByUsername(USERNAME);
+    user.setDuccAgreement(signDucc(user, 3));
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java:395]: was this change intentional?

**Review by jmthibault79 — COMMENTED** [review]
I have only reviewed simple changes so far.

Is the linked doc up-to-date in terms of what we should expect to see working?

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java [hunk 4]
```diff
@@ -139,7 +139,7 @@ default List<DbCardCount> findSurveyCounts(SearchTerm searchTerm) {
               + "from DbCriteria c "
               + "where c.standard = :standard "
               + "and c.conceptId in (:conceptIds) "
-              + "and match(c.fullText, concat('+[', :domainId, '_rank1]')) > 0")
+              + "and cast(match(c.fullText, concat('+[', :domainId, '_rank1]')) as integer) > 0")
```
Comment (tarekmamdouh) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java:142]: did the cast calls work eventually? I thought there were mysql errors

File: api/src/main/resources/application.properties [hunk 5]
```diff
@@ -29,10 +29,13 @@ spring.jpa.properties.hibernate.dialect=org.pmiops.workbench.cdr.MySQLDialect
 spring.jpa.properties.hibernate.jdbc.batch_size=100
 spring.jpa.properties.hibernate.order_inserts=true
 
+spring.security.filter.dispatcher-types=FORWARD
```
Comment (tarekmamdouh) [inline api/src/main/resources/application.properties:32]: did you remember why was this line for?

File: api/src/test/resources/application.properties [hunk 6]
```diff
@@ -9,9 +9,9 @@ spring.liquibase.enabled=false
 spring.main.allow-bean-definition-overriding=true
 
 # Avoid log spam in tests
-spring.main.log-startup-info=OFF
+spring.main.log-startup-info=false
 spring.main.banner-mode=off
-logging.level.org.springframework=WARN
+logging.level.org.springframework=DEBUG
```
Comment (tarekmamdouh) [inline api/src/test/resources/application.properties:14]: please revert this to WARN, we don't want to flood the logs

File: api/src/test/resources/application.properties [hunk 7]
```diff
@@ -24,5 +24,5 @@ spring.jpa.properties.hibernate.show_sql=false
 #spring.jpa.properties.hibernate.use_sql_comments=true
 
 # Alternative settings for debugging. When combined with those above, seems to give redundant output.
-#logging.level.org.hibernate.SQL=DEBUG
-#logging.level.org.hibernate.type.descriptor.sql.BasicBinder=TRACE
+logging.level.org.hibernate.SQL=DEBUG
```
Comment (tarekmamdouh) [inline api/src/test/resources/application.properties:27]: and comment this out as well.

File: api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java (hunk 3 above)
Comment (yonghaoy) [reply api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java:395]: yes. The previous test does not work because it tried to insert 3 DUCC for the same user without deleting existing entity. It violates a FK violation. If we still want to test all 3, we would need to add DuccDao then delete that old one. So I feel we can just test one actually from the test name. Not sure if we need to test multiple. But I do not have too much context on this, there might be some easier fix?

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java (hunk 4 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CBCriteriaDao.java:142]: it works. The issue is something else

File: api/src/main/resources/application.properties (hunk 5 above)
Comment (yonghaoy) [reply api/src/main/resources/application.properties:32]: ah, this might not needed. Deleted it

File: api/src/test/resources/application.properties (hunk 6 above)
Comment (yonghaoy) [reply api/src/test/resources/application.properties:14]: ah, okay done. (but there is unit test, not production server)

File: api/src/main/java/org/pmiops/workbench/absorb/Credentials.java (hunk 1 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/absorb/Credentials.java:3]: no.  no change needed here 👍

File: api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java (hunk 3 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java:395]: It's normal for a user to sign a version of the DUCC and then later sign a different one, so that's an important use case to test.

Can you say more about the violation that this causes?  I can work on an update to this test if I understand the problem better.

**Review by freemabd — APPROVED** [review]
LGTM. I've deployed locally and did some spot checks on Data Apps and things look good. Thanks for all the hard work on this @yonghaoy!

File: api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java (hunk 3 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java:395]: Here is the fix: https://github.com/all-of-us/workbench/pull/8527

If you approve it, I will merge it into **this** branch.

File: api/src/main/java/org/pmiops/workbench/config/WebMvcConfig.java [hunk 8]
```diff
@@ -90,4 +82,14 @@ static ServletContext getRequestServletContext() {
         .getRequest()
         .getServletContext();
   }
+
+  @Override
+  public void addCorsMappings(CorsRegistry registry) {
+    registry
+        .addMapping("/**")
+        .allowedMethods("GET", "POST", "PUT", "DELETE", "OPTIONS")
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/config/WebMvcConfig.java:90]: CorsInterceptor: `response.setHeader(METHODS_NAME, "GET, HEAD, POST, PUT, DELETE, PATCH, TRACE, OPTIONS");`

Looks like we're missing HEAD, PATCH, and TRACE here.

**Review by jmthibault79 — CHANGES_REQUESTED** [review]
[EDIT: **My local DB was not running.**  Fixing that solved the problem.  The error could be clearer, but I don't think this should block the upgrade.]

~I attempted to run a few command line tools, and failed:~

~`./project.rb update-cdr-config-local` and `./project.rb set-authority-local --email joel-local-may23@fake-research-aou.org --remove-all --dry_run` both had the same result:~

```
2024-04-30T14:21:48.442-04:00  INFO 48669 --- [           main] .t.CommandLineToolApplicationRunListener : Registering fake REQUEST scope to the context bean factory for command-line tools
2024-04-30T14:21:48.445-04:00  INFO 48669 --- [           main] o.p.w.tools.cdrconfig.UpdateCdrConfig    : Starting UpdateCdrConfig using Java 17.0.9 with PID 48669 (/Users/thibault/src/aou/api/build/classes/java/__tools__ started by thibault in /Users/thibault/src/aou/api)
2024-04-30T14:21:48.446-04:00  INFO 48669 --- [           main] o.p.w.tools.cdrconfig.UpdateCdrConfig    : No active profile set, falling back to 1 default profile: "default"
2024-04-30T14:21:48.603-04:00  INFO 48669 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
2024-04-30T14:21:48.729-04:00  INFO 48669 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 122 ms. Found 36 JPA repository interfaces.
2024-04-30T14:21:48.759-04:00  INFO 48669 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
2024-04-30T14:21:48.775-04:00  INFO 48669 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 16 ms. Found 36 JPA repository interfaces.
2024-04-30T14:21:48.902-04:00  WARN 48669 --- [           main] s.c.a.AnnotationConfigApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.pmiops.workbench.db.WorkbenchDbConfig': Unsatisfied dependency expressed through constructor parameter 0: Error creating bean with name 'org.pmiops.workbench.db.Params': Failed to instantiate [org.pmiops.workbench.db.Params$$SpringCGLIB$$0]: Constructor threw exception
```

File: api/src/main/java/org/pmiops/workbench/cdr/MySQLDialect.java [hunk 9]
```diff
@@ -1,30 +1,43 @@
...
     // <concept>.vocabulary_id,
     // <concept>.synonyms) against (<string> in boolean mode)". Clients must pass in each of the
     // corresponding fields -- conceptName, conceptCode, vocabularyId, synonymsStr --
     // to make JQL alias resolution work properly. Example:
     // matchConcept(c.conceptName, c.conceptCode, c.vocabularyId, c.synonymsStr, ?1) > 0
-    registerFunction(
-        "matchConcept",
-        new SQLFunctionTemplate(
-            StandardBasicTypes.DOUBLE, "match(?1, ?2, ?3, ?4) against (?5 in boolean mode)"));
+    functionContributions
```
[the remark this answers is not in the document]
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/cdr/MySQLDialect.java:29]: restored to the original version from main branch

**Comment by yonghaoy** [comment]
> [EDIT: **My local DB was not running.** Fixing that solved the problem. The error could be clearer, but I don't think this should block the upgrade.]
> 
> ~I attempted to run a few command line tools, and failed:~
> 
> ~`./project.rb update-cdr-config-local` and `./project.rb set-authority-local --email joel-local-may23@fake-research-aou.org --remove-all --dry_run` both had the same result:~
> 
> ```
> 2024-04-30T14:21:48.442-04:00  INFO 48669 --- [           main] .t.CommandLineToolApplicationRunListener : Registering fake REQUEST scope to the context bean factory for command-line tools
> 2024-04-30T14:21:48.445-04:00  INFO 48669 --- [           main] o.p.w.tools.cdrconfig.UpdateCdrConfig    : Starting UpdateCdrConfig using Java 17.0.9 with PID 48669 (/Users/thibault/src/aou/api/build/classes/java/__tools__ started by thibault in /Users/thibault/src/aou/api)
> 2024-04-30T14:21:48.446-04:00  INFO 48669 --- [           main] o.p.w.tools.cdrconfig.UpdateCdrConfig    : No active profile set, falling back to 1 default profile: "default"
> 2024-04-30T14:21:48.603-04:00  INFO 48669 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
> 2024-04-30T14:21:48.729-04:00  INFO 48669 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 122 ms. Found 36 JPA repository interfaces.
> 2024-04-30T14:21:48.759-04:00  INFO 48669 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
> 2024-04-30T14:21:48.775-04:00  INFO 48669 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 16 ms. Found 36 JPA repository interfaces.
> 2024-04-30T14:21:48.902-04:00  WARN 48669 --- [           main] s.c.a.AnnotationConfigApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.pmiops.workbench.db.WorkbenchDbConfig': Unsatisfied dependency expressed through constructor parameter 0: Error creating bean with name 'org.pmiops.workbench.db.Params': Failed to instantiate [org.pmiops.workbench.db.Params$$SpringCGLIB$$0]: Constructor threw exception
> ```

I am able to run this command when having a running local API server and get the same error if I ran without local API, I assume my local mysql would start only when I have api running?

**Review by jmthibault79 — APPROVED** [review]
Thanks for the updates.

Looks good, no apparent problems when running locally.

**Review by tarekmamdouh — APPROVED** [review]
LGTM, tested locally with no issues that I could observe.

File: api/src/main/java/org/pmiops/workbench/absorb/Credentials.java (hunk 1 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/absorb/Credentials.java:3]: Converted jetbrains to jakarta in #8537
